scio-extra/src/main/scala/com/spotify/scio/extra/sparkey/SparkeyUri.scala

```diff
@@ -19,104 +19,138 @@ package com.spotify.scio.extra.sparkey
 
+import java.nio.file.Path
+import scala.collection.mutable
 import scala.jdk.CollectionConverters._
 
-/**
- * Represents the base URI for a Sparkey index and log file, either on the local or a remote file
- * system. For remote file systems, `basePath` should be in the form
- * 'scheme://<bucket>/<path>/<sparkey-prefix>'. For local files, it should be in the form
- * '/<path>/<sparkey-prefix>'. Note that `basePath` must not be a folder or GCS bucket as it is a
- * base path representing two files - <sparkey-prefix>.spi and <sparkey-prefix>.spl.
- */
-trait SparkeyUri extends Serializable {
-  val basePath: String
-  def getReader: SparkeyReader
-
-  private[sparkey] def exists: Boolean
-  override def toString: String = basePath
-}
+case class InvalidShards(str: String) extends RuntimeException(str)
 
-private[sparkey] object SparkeyUri {
-  def apply(basePath: String, opts: PipelineOptions): SparkeyUri =
-    if (ScioUtil.isLocalUri(new URI(basePath))) {
-      LocalSparkeyUri(basePath)
-    } else {
-      RemoteSparkeyUri(basePath, opts)
-    }
+object SparkeyUri {
   def extensions: Seq[String] = Seq(".spi", ".spl")
-
-  implicit def coderSparkeyURI: Coder[SparkeyUri] = Coder.kryo[SparkeyUri]
+  def basePathForShard(basePath: String, shardIndex: Short, numShards: Short): String =
+    f"$basePath/part-$shardIndex%05d-of-$numShards%05d"
 }
 
-private case class LocalSparkeyUri(basePath: String) extends SparkeyUri {
-  override def getReader: SparkeyReader =
-    new ThreadLocalSparkeyReader(new File(basePath))
-  override private[sparkey] def exists: Boolean =
-    SparkeyUri.extensions.map(e => new File(basePath + e)).exists(_.exists)
-}
+case class SparkeyUri(path: String) {
+  private[sparkey] val isLocal = ScioUtil.isLocalUri(new URI(path))
+  private val isSharded = path.endsWith("*")
+  private[sparkey] val basePath =
+    if (!isSharded) path else path.split("/").dropRight(1).mkString("/")
 
-private object RemoteSparkeyUri {
-  def apply(basePath: String, options: PipelineOptions): RemoteSparkeyUri =
-    RemoteSparkeyUri(basePath, RemoteFileUtil.create(options))
-}
+  private[sparkey] def globExpression: String = s"$basePath/part-*"
 
-private case class RemoteSparkeyUri(basePath: String, rfu: RemoteFileUtil) extends SparkeyUri {
-  override def getReader: SparkeyReader = {
-    val uris = SparkeyUri.extensions.map(e => new URI(basePath + e))
-    val paths = rfu.download(uris.asJava).asScala
-    new ThreadLocalSparkeyReader(paths.head.toFile)
+  private[sparkey] def sparkeyUriForShard(shard: Short, numShards: Short): SparkeyUri =
+    SparkeyUri(SparkeyUri.basePathForShard(basePath, shard, numShards))
+
+  private[sparkey] def paths: Seq[ResourceId] = {
+    if (isSharded) throw new IllegalStateException("Internal use only for single files.")
+    SparkeyUri.extensions.map { e =>
+      val url = if (isLocal) new File(basePath + e).toString else new URI(basePath + e).toString
```

Comment:
Why are we creating a `File` or `URI` and converting it straight to string ? Won't this get back the original string given in the constructor ?

Reply:
also, can't we use URI for all cases (local & remote) ?